src/java.base/share/classes/java/lang/Class.java

```diff
@@ -593,14 +598,80 @@ public boolean isInlineClass() {
      *         inline class; an empty {@link Optional} otherwise
      * @since Valhalla
      */
-    public Optional<Class<T>> referenceType() {
-        return valType != null ? Optional.ofNullable(refType) : Optional.of(this);
+    public Optional<Class<?>> referenceType() {
+        if (isPrimitive() || isInterface() || isArray())
+            return Optional.empty();
+        ensureProjectionTypesInited();
+        System.out.println(getName() + " " + Arrays.toString(projectionTypes));
+        return projectionTypes.length == 2 ? Optional.of(projectionTypes[1]) : Optional.empty();
     }
 
     // set by VM if this class is an inline type
     // otherwise, these two fields are null
     private transient Class<T> valType;
     private transient Class<T> refType;
+    private transient Class<?>[] projectionTypes;
+
+    private synchronized void ensureProjectionTypesInited() {
+        if (isPrimitive() || isArray() || isInterface())
+            return;
+
+        if (projectionTypes != null)
+            return;
+
+        if (isInlineClass()) {
+            Class<?> superClass = getSuperclass();
+            System.out.println("superclass " + superClass);
+            if (superClass != Object.class && superClass.isReferenceProjectionType()) {
+                projectionTypes = new Class<?>[] { this, superClass };
+            } else {
+                projectionTypes = new Class<?>[] { this };
+            }
+        } else if (isReferenceProjectionType()) {
+            projectionTypes = new Class<?>[] { valueProjectionType(), this };
+        } else {
+            projectionTypes = EMPTY_CLASS_ARRAY;
+        }
+    }
+
+    private boolean isReferenceProjectionType() {
+        if (isPrimitive() || isInterface() || isArray() || isInlineClass())
+            return false;
+
+        if (projectionTypes != null) {
+            return projectionTypes.length == 2 && projectionTypes[1] == this;
+        }
+
+        int mods = getModifiers();
+        if (!Modifier.isAbstract(mods)) {
+            return false;
+        }
+
+        return valueProjectionType() != null;
```

Comment:
This path computed the valueProjecttionType and then discards it to be computed again on line 631.

Reply:
I don't think this path is performance critical.  We can look into this if performance turns out to be an issue.